mica/stats/guide_stats.py

```diff
@@ -0,0 +1,570 @@
+    catalog.sort('idx')
+    # Filter the catalog to be just guide stars
+    catalog = catalog[(catalog['type'] == 'GUI') | (catalog['type'] == 'BOT')]
+    # Get the position deltas relative to onboard solution
+    dy, dz, star_info, yag, zag = _deltas_vs_obc_quat(eng_data, times, catalog)
+    # And add the deltas to the table
+    for slot in range(0, 8):
+        if slot not in dy:
+            continue
+        eng_data['dy{}'.format(slot)] = dy[slot].data
+        eng_data['dz{}'.format(slot)] = dz[slot].data
+        eng_data['cat_yag{}'.format(slot)] = yag[slot]
+        eng_data['cat_zag{}'.format(slot)] = zag[slot]
+        cat_entry = catalog[catalog['slot'] == slot][0]
+        dmag = eng_data['AOACMAG{}'.format(slot)] - cat_entry['mag']
+        eng_data['dmag'] = dmag.data
+    return eng_data, times, star_info
+
+
+def consecutive(data, stepsize=1):
+        return np.split(data, np.where(np.diff(data) != stepsize)[0] + 1)
+
+
+def calc_gui_stats(start, stop, data, times, star_info):
+    logger.info("calculating statistics")
+    gui_stats = {}
+    for slot in range(0, 8):
+        if 'dy{}'.format(slot) not in data.colnames:
+            continue
+        stats = {}
+        aoacfct = data['AOACFCT{}'.format(slot)]
+        stats['n_samples'] = len(aoacfct)
+        if len(aoacfct) == 0:
+            gui_stats[slot] = stats
+            continue
+        stats['n_track'] = np.count_nonzero(aoacfct == 'TRAK')
+        stats['f_track'] = stats['n_track'] / stats['n_samples']
+        stats['f_racq'] = np.count_nonzero(aoacfct == 'RACQ') / stats['n_samples']
+        stats['f_srch'] = np.count_nonzero(aoacfct == 'SRCH') / stats['n_samples']
+        stats['f_none'] = np.count_nonzero(aoacfct == 'NONE') / stats['n_samples']
+        if np.all(aoacfct != 'TRAK'):
+            gui_stats[slot] = stats
+            continue
+
+        trak = data[aoacfct == 'TRAK']
+        ok_flags = ((trak['AOACIIR{}'.format(slot)] == 'OK ')
+                    & (trak['AOACISP{}'.format(slot)] == 'OK '))
+        stats['n_kalman'] = np.count_nonzero(ok_flags)
+        stats['no_track'] = (stats['n_samples'] - stats['n_track']) / stats['n_samples']
+        stats['f_obc_bad'] = (stats['n_track'] - stats['n_kalman']) / stats['n_track']
+        stats['f_common_col'] = np.count_nonzero(trak['AOACICC{}'.format(slot)] == 'ERR') / stats['n_track']
+        stats['f_sat_pix'] = np.count_nonzero(trak['AOACISP{}'.format(slot)] == 'ERR') / stats['n_track']
+        stats['f_def_pix'] = np.count_nonzero(trak['AOACIDP{}'.format(slot)] == 'ERR') / stats['n_track']
+        stats['f_ion_rad'] = np.count_nonzero(trak['AOACIIR{}'.format(slot)] == 'ERR') / stats['n_track']
+        stats['f_mult_star'] = np.count_nonzero(trak['AOACIMS{}'.format(slot)] == 'ERR') / stats['n_track']
+        stats['f_quad_bound'] = np.count_nonzero(trak['AOACIQB{}'.format(slot)] == 'ERR') / stats['n_track']
+
+        track_interv = consecutive(np.flatnonzero(
```

Comment:
This is fine, but don't forget about the existing utilities `logical_intervals` and `state_intervals` that do roughly the same thing (http://cxc.cfa.harvard.edu/mta/ASPECT/tool_doc/eng_archive/utils.html).